packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts

```diff
@@ -0,0 +1,567 @@
+          this.current,
+        ),
+      );
+    } else {
+      this.tokens.push(
+        SyntaxToken.create(
+          SyntaxTokenKind.STRING_LITERAL,
+          this.start - 3,
+          this.current - this.start + 6,
+          string,
+        ),
+      );
+      this.skip(3);
+    }
+  }
+
+  functionExpression() {
+    this.skip();
+    let string = '';
+    while (!this.isAtEnd() && !(this.peek() === '\n') && !(this.peek() === '`')) {
+      if (this.peek() === '\\') {
+        string += this.escapedString();
+      } else {
+        string += this.advance();
+      }
+    }
+
+    if (this.isAtEnd() || this.peek() === '\n') {
+      this.errors.push(
+        new ParsingError(
+          ParsingErrorCode.INVALID,
+          'Unclosed function expression',
+          this.start - 1,
+          this.current,
+        ),
+      );
+    } else {
+      this.tokens.push(
+        SyntaxToken.create(
+          SyntaxTokenKind.FUNCTION_EXPRESSION,
+          this.start - 1,
+          this.current - this.start + 2,
+          string,
+        ),
+      );
+      this.advance();
+    }
+  }
+
+  quotedVariable() {
+    this.skip();
+    let string = '';
+    while (!this.isAtEnd() && !(this.peek() === '\n') && !(this.peek() === '"')) {
+      if (this.peek() === '\\') {
+        string += this.escapedString();
+      } else {
+        string += this.advance();
+      }
+    }
+
+    if (this.isAtEnd() || this.peek() === '\n') {
+      this.errors.push(
+        new ParsingError(
+          ParsingErrorCode.INVALID,
+          'Unclosed quoted variable',
+          this.start - 1,
+          this.current,
+        ),
+      );
+    } else {
+      this.tokens.push(
+        SyntaxToken.create(
+          SyntaxTokenKind.QUOTED_STRING,
+          this.start - 1,
+          this.current - this.start + 2,
+          string,
+        ),
+      );
+      this.advance();
+    }
+  }
+
+  identifier() {
+    const first = this.peek();
+    if (!first || !isAlpha(first)) {
+      return;
+    }
+    this.advance();
+    let c = this.peek();
+    while (c && isAlphaNumeric(c)) {
+      this.advance();
+      c = this.peek();
+    }
+    this.addToken(SyntaxTokenKind.IDENTIFIER);
+  }
+
+  numericLiteral() {
+    let isFloat = false;
+    const first = this.peek();
+    if (!first || !isDigit(first)) {
+      return;
+    }
+    this.advance();
+    let c = this.peek();
+    while (c && (isDigit(c) || c === '.')) {
+      if (c === '.') {
+        if (isFloat) {
+          break;
+        }
+        isFloat = true;
+      }
+      this.advance();
+      c = this.peek();
+    }
+
+    if (c && (isAlpha(c) || c === '.')) {
+      while (c && (isAlphaNumeric(c) || c === '.')) {
+        this.advance();
+        c = this.peek();
+      }
+      this.errors.push(
+        new ParsingError(ParsingErrorCode.INVALID, 'Invalid number', this.start, this.current),
+      );
+    } else {
+      this.addToken(SyntaxTokenKind.NUMERIC_LITERAL);
+    }
+  }
```

Comment:
Why is this method so complex than it is in AML @huy-dna?

Reply:
This one also checks for adjacent alphabetical characters after the digits. If there are such characters, the number is marked invalid.
In AML, the lexer doesn't check this, which leads to the problem mentioned in this:
https://app.asana.com/0/1152839153201473/1205071037227754/f

Reply:
Oh I see. But I don't think the lexer should care about it. IMO, this logic should be in the parser. TypeScript is still able to scan this one `2const` as two separate tokens `2` and `const`